gms/gossiper.hh

```diff
@@ -84,6 +85,45 @@ class i_endpoint_state_change_subscriber;
 
 class feature_service;
 
+class gossip_query_token_status_response {
+public:
+    // The keys can be:
+    // - nr_known_nodes : Number of nodes this node knows about
+    // - status_token_hash: The hash of the gossip status and tokens for nodes
+    // - normal_token_hash: The hash of the tokens for nodes in NORMAL status
+    // - bootstrap_token_hash: The hash of the tokens for nodes in BOOTSTRAPPING status
+    // - leaving_nodes_hash: The hash of the nodes that are leaving the cluster
+    // - pending_ranges_hash: The hash of pending ranges
+    // - same_on_all_shards: If all the shards have the same hashes
+    static constexpr const char* nr_known_nodes = "nr_known_nodes";
+    static constexpr const char* status_token_hash = "status_token_hash";
+    static constexpr const char* normal_token_hash = "normal_token_hash";
+    static constexpr const char* bootstrap_token_hash = "bootstrap_token_hash";
+    static constexpr const char* leaving_nodes_hash = "leaving_nodes_hash";
+    static constexpr const char* pending_ranges_hash = "pending_ranges_hash";
+    static constexpr const char* same_on_all_shards = "same_on_all_shards";
```

Comment:
Why are we using ` map<sstring, sstring>` and losing type safety? Since we're defining a class, it would be better to just define a bunch of fields: `int8_t nr_known_nodes`, `uint64_t status_token_hash` etc.

Reply:
Image later we need a new item in the response, how do we extend it without breaking the rpc compatibility? 

gossip_query_token_status_response as the transport, does not care the types of the items, all of them are serialized to string.

Reply:
RPC doesn't care what "struct/class" you're using. If it can be implemented in an extensible way with `map<string, string>`, then it can be implemented using a bunch of type-safe fields. You just need to use a proper serializer.

 > gossip_query_token_status_response as the transport, does not care the types of the items, all of them are serialized to string.

Then I wonder why we need all these different names at all. Can't we just have a single hash?

Reply:
Given
```
struct resposne {
       int key1,
       int key2;
}
future<response> rpc_call(int x, int y, rpc::optional<z>)
```
Now, we need to add key3. How do we extend the response in a compatible way?

We can do:
```
struct resposne {
       int key1,
       int key2;
}
future<response, rpc::optional<key3>> rpc_call(int x, int y, rpc::optional<z>)

```
This is depreciated also ugly. 

We can use the versions in the idl, but it adds extra complexity. It is also not widely used.